src/transformState.js

```diff
@@ -3,9 +3,39 @@
 /**
  * @param {Object} state
  * @param {Object[]} actions
- */
+*/
+
 function transformState(state, actions) {
-  // write code here
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(state, action.extraData);
+        break;
+
+      case 'removeProperties':
+        removeSpecificKeys(action);
+        break;
+
+      case 'clear':
+        removeAllKeys();
+        break;
+
+      default:
+        throw new Error(`No ${action.type} action type found`);
+    }
+  }
+
+  function removeAllKeys() {
+    for (const key in state) {
+      delete state[key];
+    }
+  }
+
+  function removeSpecificKeys(arrayToClean) {
+    for (const key of arrayToClean.keysToRemove) {
+      delete state[key];
+    }
+  }
```

Comment:
Would be better to move this functions outside the `transformState` function. Since these functions are now created every time the `transformState` function is called